wideq/device/__init__.py

```diff
@@ -0,0 +1,215 @@
+        """Look up the encoded value for a friendly enum name.
+        """
+        options = self.value(key).options
+        options_inv = {v: k for k, v in options.items()}  # Invert the map.
+        return options_inv[name]
+
+    def enum_name(self, key, value):
+        """Look up the friendly enum name for an encoded value.
+        """
+        options = self.value(key).options
+        return options[value]
+
+    @property
+    def binary_monitor_data(self):
+        """Check that type of monitoring is BINARY(BYTE).
+        """
+        return self.data['Monitoring']['type'] == 'BINARY(BYTE)'
+
+    def decode_monitor_binary(self, data):
+        """Decode binary encoded status data.
+        """
+        decoded = {}
+        for item in self.data['Monitoring']['protocol']:
+            key = item['value']
+            value = 0
+            for v in data[item['startByte']:item['startByte'] +
+                          item['length']]:
+                value = (value << 8) + v
+            decoded[key] = str(value)
+        return decoded
+
+    def decode_monitor_json(self, data):
+        """Decode a bytestring that encodes JSON status data."""
+        return json.loads(data.decode('utf8'))
+
+    def decode_monitor(self, data):
+        """Decode  status data."""
+        if self.binary_monitor_data:
+            return self.decode_monitor_binary(data)
+        else:
+            return self.decode_monitor_json(data)
+
+
+class Device(object):
+    """A higher-level interface to a specific device.
+
+    Unlike `DeviceInfo`, which just stores data *about* a device,
+    `Device` objects refer to their client and can perform operations
+    regarding the device.
+    """
+
+    def __init__(self, client, device):
+        """Create a wrapper for a `DeviceInfo` object associated with a
+        `Client`.
+        """
+
+        self.client = client
+        self.device = device
+        self.model = client.model_info(device)
+
+    def _set_control(self, key, value):
+        """Set a device's control for `key` to `value`.
+        """
+
+        self.client.session.set_device_controls(
+            self.device.id,
+            {key: value},
+        )
+
+    def _get_config(self, key):
+        """Look up a device's configuration for a given value.
+
+        The response is parsed as base64-encoded JSON.
+        """
+
+        data = self.client.session.get_device_config(
+            self.device.id,
+            key,
+        )
+        return json.loads(base64.b64decode(data).decode('utf8'))
+
+    def _get_control(self, key):
+        """Look up a device's control value.
+        """
+
+        data = self.client.session.get_device_config(
+            self.device.id,
+            key,
+            'Control',
+        )
+
+        # The response comes in a funky key/value format: "(key:value)".
+        _, value = data[1:-1].split(':')
+        return value
+
+    def monitor_start(self):
```

Comment:
This is the only code change, moved `monitor_start` and `monitor_end` into the base Device class.  Looking at your `ACDevice` and wkd8176's fork with washer/dryer/etc. devices, these 2 methods are always identical, so it makes more sense to put this in the base class.